File: frontend/src/app/core/services/technicien-profile.ts
import { Injectable, signal } from "@angular/core";
import { HttpClient } from "@angular/common/http";
import { Observable, tap } from "rxjs";
import { environment } from "../../../environments/environment";
export interface TechnicienProfileData {
  // Informations personnelles
  cin: string;
  dateNaissance: string;
  sexe: string;
  adresse: string;
  ville: string;
  gouvernorat: string;
  telephoneSecondaire: string;
  // Informations professionnelles
  matricule: string;
  metier: string;
  specialite: string;
  niveauEtudes: string;
  diplome: string;
  etablissement: string;
  anneeObtention: number | null;
  anneesExperience: number | null;
  entrepriseActuelle: string;
  statutProfessionnel: string;
  // Compétences
  competenceCctv: boolean;
  competenceControleAcces: boolean;
  competenceReseaux: boolean;
  competenceFibreOptique: boolean;
  competenceElectricite: boolean;
  competenceDomotique: boolean;
  competenceMaintenanceInformatique: boolean;
  competenceAlarmesIntrusion: boolean;
  competenceDetectionIncendie: boolean;
  autresCompetences: string;
  // Sécurité électronique / CCTV
  travailleSecuriteElectronique: boolean;
  experienceCamerasIp: boolean;
  experienceNvrDvr: boolean;
  marquesMaitrisees: string;
  configurationReseau: boolean;
  pointageBiometrique: boolean;
  interphonieIp: boolean;
  soudureFibreOptique: boolean;
  certifications: string;
  // Permis et mobilité
  permisConduire: boolean;
  categoriePermis: string;
  vehiculePersonnel: boolean;
  disponibleDeplacement: boolean;
  zonesIntervention: string;
  // Disponibilité
  disponibleImmediatement: boolean;
  typeDisponibilite: string;
  salaireSouhaite: number | null;
  // Bancaire
  banque: string;
  rib: string;
  // Observations
  commentaires: string;
}
export interface TechnicienDocuments {
  cinDocument?: File | null;
  cvDocument?: File | null;
  diplomeDocument?: File | null;
  attestationTravail?: File | null;
  photoIdentite?: File | null;
  casierJudiciaire?: File | null;
}
@Injectable({ providedIn: "root" })
export class TechnicienProfileService {
  private readonly apiUrl = `${environment.apiUrl}/technicien/profil`;
  hasProfile = signal(false);
  constructor(private http: HttpClient) {}
  checkExists(): Observable<boolean> {
    return this.http.get<boolean>(`${this.apiUrl}/me/exists`).pipe(
      tap((exists) => this.hasProfile.set(exists))
    );
  }
  getMyProfile(): Observable<any> {
    return this.http.get<any>(`${this.apiUrl}/me`);
  }
  saveProfile(data: TechnicienProfileData, documents: TechnicienDocuments): Observable<any> {
    const formData = new FormData();
    formData.append("data", new Blob([JSON.stringify(data)], { type: "application/json" }));
    if (documents.cinDocument) formData.append("cinDocument", documents.cinDocument);
    if (documents.cvDocument) formData.append("cvDocument", documents.cvDocument);
    if (documents.diplomeDocument) formData.append("diplomeDocument", documents.diplomeDocument);
    if (documents.attestationTravail) formData.append("attestationTravail", documents.attestationTravail);
    if (documents.photoIdentite) formData.append("photoIdentite", documents.photoIdentite);
    if (documents.casierJudiciaire) formData.append("casierJudiciaire", documents.casierJudiciaire);
    return this.http.post(this.apiUrl, formData).pipe(
      tap(() => this.hasProfile.set(true))
    );
  }
  getDocumentUrl(type: "CIN" | "CV" | "DIPLOME" | "ATTESTATION" | "PHOTO" | "CASIER"): string {
    return `${this.apiUrl}/me/documents/${type}`;
  }
}

File: backend/src/main/java/com/heyaatounisiya/backend/entity/TechnicienProfile.java
package com.heyaatounisiya.backend.entity;

import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "technicien_profiles")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
@Builder
public class TechnicienProfile {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @OneToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "user_id", nullable = false, unique = true)
    private User user;

    // ===== Informations personnelles =====
    private String cin;
    private LocalDate dateNaissance;

    @Enumerated(EnumType.STRING)
    private Sexe sexe;

    @Column(length = 500)
    private String adresse;

    private String ville;
    private String gouvernorat;
    private String telephoneSecondaire;

    // ===== Informations professionnelles =====
    private String matricule;

    @Enumerated(EnumType.STRING)
    private Metier metier;

    private String specialite;

    @Enumerated(EnumType.STRING)
    private NiveauEtudes niveauEtudes;

    private String diplome;
    private String etablissement;
    private Integer anneeObtention;
    private Integer anneesExperience;
    private String entrepriseActuelle;

    @Enumerated(EnumType.STRING)
    private StatutProfessionnel statutProfessionnel;

    // ===== Compétences techniques =====
    private boolean competenceCctv;
    private boolean competenceControleAcces;
    private boolean competenceReseaux;
    private boolean competenceFibreOptique;
    private boolean competenceElectricite;
    private boolean competenceDomotique;
    private boolean competenceMaintenanceInformatique;
    private boolean competenceAlarmesIntrusion;
    private boolean competenceDetectionIncendie;

    @Column(length = 1000)
    private String autresCompetences;

    // ===== Spécifique sécurité électronique (CCTV) =====
    private boolean travailleSecuriteElectronique;
    private boolean experienceCamerasIp;
    private boolean experienceNvrDvr;

    @Column(length = 500)
    private String marquesMaitrisees;

    private boolean configurationReseau;
    private boolean pointageBiometrique;
    private boolean interphonieIp;
    private boolean soudureFibreOptique;

    @Column(length = 500)
    private String certifications;

    // ===== Permis et mobilité =====
    private boolean permisConduire;

    @Enumerated(EnumType.STRING)
    private CategoriePermis categoriePermis;

    private boolean vehiculePersonnel;
    private boolean disponibleDeplacement;

    @Column(length = 1000)
    private String zonesIntervention;

    // ===== Documents (chemins relatifs vers les fichiers stockés) =====
    private String cinDocumentPath;
    private String cvDocumentPath;
    private String diplomeDocumentPath;
    private String attestationTravailPath;
    private String photoIdentitePath;
    private String casierJudiciairePath;

    // ===== Disponibilité =====
    private boolean disponibleImmediatement;

    @Enumerated(EnumType.STRING)
    private TypeDisponibilite typeDisponibilite;

    private Double salaireSouhaite;

    // ===== Coordonnées bancaires (optionnel) =====
    private String banque;
    private String rib;

    // ===== Observations =====
    @Column(length = 2000)
    private String commentaires;

    @Column(nullable = false, updatable = false)
    private LocalDateTime createdAt;

    private LocalDateTime updatedAt;

    @PrePersist
    protected void onCreate() {
        this.createdAt = LocalDateTime.now();
        this.updatedAt = LocalDateTime.now();
    }

    @PreUpdate
    protected void onUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}
